books/projects/x86isa/machine/instructions/bit.lisp

```diff
@@ -777,4 +777,46 @@
                (x86 (write-*ip proc-mode temp-rip x86)))
               x86))
 
+(defmacro bswap (n src)
+  (if (equal n 0)
+    0
+    `(logapp 8 (logtail ,(* (1- n) 8) ,src)
+             (bswap ,(1- n) ,src))))
+
+(def-inst x86-bswap
+
+  ;; 0F C8 + rd: BSWAP r32
+  ;; 0F C8 + rd: BSWAP r64
+
+  :parents (two-byte-opcodes)
+
+  :returns (x86 x86p :hyp (x86p x86))
+  :guard-hints (("Goal" :in-theory (disable reg-indexp unsigned-byte-p)))
+
+  :body
+
+  (b* (;; Only 32 and 64 bit registers are allowed
+       ((the (integer 2 8) operand-size)
+        (select-operand-size
+         proc-mode nil rex-byte nil prefixes nil nil nil x86))
+
+       ;; SDM says undefined behavior for 16-bit register
+       ((when (equal operand-size 2)) (!!ms-fresh :bswap-16-bit))
```

Comment:
The Intel manual says that the result is undefined in this case. I believe that, ideally, we should handle this similarly to how undefined flags are handled after certain operations. But maybe for now we could add a comment saying all of this. Would you be willing to add that comment?

Reply:
Using the mechanism used for undefined flags etc. is definitely the proper way to do it, but I am not very familiar with that mechanism. I'll look into it and use that if it isn't too difficult. Otherwise, I'll add the comment.